Data/HashMap/Base.hs

```diff
@@ -1822,6 +1799,56 @@ updateOrConcatWithKey f ary1 ary2 = A.run $ do
     return mary
 {-# INLINABLE updateOrConcatWithKey #-}
 
+-- Helper functions for very rare cases.
+
+-- Remove (any) one element from an array of hashmaps
+unConsA :: A.Array (HashMap k v) -> UnCons k v
+unConsA ary = case A.length ary of
+  0 -> WasEmpty
+  1 -> case A.index ary 0 of
+    Empty -> WasEmpty -- That would be weird, but fine.
+    Leaf _ l -> NowEmpty l
+    BitmapIndexed bm ary' -> case unConsA ary' of
+      WasEmpty -> WasEmpty -- That would be weird, but fine.
+      NowEmpty l -> NowEmpty l
+      UnConsed l a' -> UnConsed l (A.singleton (BitmapIndexed bm a'))
+    Full ary' -> case unConsA ary' of
+        WasEmpty -> WasEmpty -- That would be weird, but fine.
+        NowEmpty l -> NowEmpty l
+        UnConsed l a' -> UnConsed l (A.singleton (Full a')) -- TODO 'Full' seems wrong
+    Collision h l1 l2 hm -> UnConsed l1 $ A.singleton $ case unConsHM hm of
+      Nothing -> Leaf h l2
+      Just (l3, hm') -> Collision h l2 l3 hm'
+  _ -> goA 0
+  where
+    goA ix = case unConsHM $ A.index ary ix of
+      Nothing -> goA $ ix + 1 -- TODO this is probably problematic when the entire array only houses Empty hashmaps
+      Just (l, hm') -> case hm' of
+        Empty -> UnConsed l $ A.delete ary ix
+        _ -> UnConsed l $ A.update ary ix hm'
+
+data UnCons k v
+  = WasEmpty
+  | NowEmpty (Leaf k v)
+  | UnConsed (Leaf k v) (A.Array (HashMap k v))
+
+-- Remove (any) one element from a hashmap
+unConsHM :: HashMap k v -> Maybe (Leaf k v, HashMap k v)
+unConsHM hm = case hm of
+  Empty -> Nothing
+  Leaf _ l -> Just (l, Empty)
+  BitmapIndexed bm ary' -> case unConsA ary' of
+    WasEmpty -> Nothing
+    NowEmpty l -> Just (l, Empty)
+    UnConsed l a' -> Just (l, BitmapIndexed bm a')
+  Full ary' -> case unConsA ary' of
+    WasEmpty -> Nothing -- That would be weird, but fine.
+    NowEmpty l -> Just (l, Empty)
+    UnConsed l a' -> Just (l, Full a') -- TODO This 'Full' seems wrong
```

Comment:
This needs careful review.

Reply:
*Both* array cases look deeply suspicious. When a map in a `BitmapIndexed` node becomes empty and is removed, we surely need to update the bitmap. I believe we also want a special case when the array has one remaining element and that element is a leaf: we should surely convert the whole thing to a leaf then. I *believe* we generally maintain the invariant that an array has at least two elements. When a map in a `Full` node goes empty and is removed, we surely need to convert to a `BitmapIndexed` node.

Reply:
@treeowl I'm not exactly clear on how the bitmap is supposed to work, but what you say makes complete sense to me.

Reply:
Well then, I think you'll have to try to figure that out. You may find it helpful to look at what `delete` does.